### PR TITLE
Elligator2 hash-to-curve for Bandersnatch

### PR DESCRIPTION
**Pull request by drskalman**
## Description

Implementation Elligator2 hash-to-curve parameters for Bandersnatch curve  according to IETF standard.

### CODE REVIEW

File: curves/ed_on_bls12_381_bandersnatch/Cargo.toml [hunk 1]
```diff
@@ -22,6 +22,7 @@ ark-relations = { version = "0.4.0", default-features = false }
 ark-serialize = { version = "0.4.0", default-features = false }
 ark-algebra-test-templates = { version = "0.4.0", default-features = false }
 ark-curve-constraint-tests = { path = "../curve-constraint-tests", default-features = false }
+sha2 = { version = "0.10", default-features = false }
```
Comment (burdges) [inline curves/ed_on_bls12_381_bandersnatch/Cargo.toml:25]: Can we avoid propogating this sha2 xmd garbage?

File: curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs [hunk 2]
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
...
+        HashToCurve,
+    };
+    use ark_ff::field_hashers::DefaultFieldHasher;
+    use sha2::Sha512;
+
+    #[test]
+    fn test_elligtor2_hash2curve_hashes_to_curve() {
+        let test_elligator2_to_curve_hasher = MapToCurveBasedHasher::<
+            Projective<BandersnatchConfig>,
+            DefaultFieldHasher<Sha512, 128>,
```
Comment (burdges) [inline curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:180]: Again we should avoid sha2 xmd for new curves.

Comment (mmagician) [reply curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:180]: this is just a test though - what is the alternative here @burdges ?

Comment (burdges) [reply curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:180]: Yeah no worries.  We should make the shake h2f stuff work eventually though.

File: curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs [hunk 3]
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
...
+    use super::*;
+    use ark_ec::hashing::{
+        curve_maps::elligator2::Elligator2Map, map_to_curve_hasher::MapToCurveBasedHasher,
+        HashToCurve,
+    };
+    use ark_ff::field_hashers::DefaultFieldHasher;
+    use sha2::Sha512;
+
+    #[test]
+    fn test_elligtor2_hash2curve_hashes_to_curve() {
```
Comment (mmagician) [inline curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:177]: I think for now it's ok to have these individual test. We should add some test templates though for hash-to-curve, I'll create an issue. Then we can have proper tests for Bandersnatch here, but also BLS12-377 & 381 on SWU.

File: curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs [hunk 4]
```diff
@@ -137,3 +138,55 @@ impl SWCurveConfig for BandersnatchConfig {
     /// generators
     const GENERATOR: SWAffine = SWAffine::new_unchecked(SW_GENERATOR_X, SW_GENERATOR_Y);
 }
+
+// Elligator hash to curve Bandersnatch
+// sage: find_z_ell2(GF(52435875175126190479447740508185965837690552500527637822603658699938581184513))
```
[the remark this answers is not in the document]
Comment (drskalman) [reply curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:143]: https://github.com/arkworks-rs/algebra/pull/758/commits/c7d149386c8551ec687279535cbb48bab6e98258

[the remark this answers is not in the document]
Comment (mmagician) [reply curves/ed_on_bls12_381_bandersnatch/src/curves/mod.rs:143]: Thank you!

File: CHANGELOG.md [hunk 5]
```diff
@@ -2,6 +2,8 @@
 
 ## Pending
 
+- [\#XXX](https://github.com/arkworks-rs/algebra/pull/XXX) Implement Elligator2 hash-to-curve parameters for Bandersnatch.
```
Comment (mmagician) [inline CHANGELOG.md:5]:
```suggestion
- [\#XXX](https://github.com/arkworks-rs/algebra/pull/758) Implement Elligator2 hash-to-curve parameters for Bandersnatch.
```

**Comment by mmagician** [comment]
Looks good. We can merge once the CHANGELOG is updated. It seems I can't commit to your branch directly.

**Comment by drskalman** [comment]
> Looks good. We can merge once the CHANGELOG is updated. It seems I can't commit to your branch directly.

I had moved it to features before but I had forgotten  to push. I also added you as a contributor to the w3f fork for future. :pray:

**Review by mmagician — APPROVED** [review]

**Pull request merged**